### PR TITLE
[Core] Improve the performance of Vector2 methods by adding _FORCE_INLINE_

### PR DESCRIPTION
**Pull request by Nazarwadim**
Added `_FORCE_INLINE_` to methods, which improved their performance by about 25%

### CODE REVIEW

**Comment by aaronfranke** [comment]
Do you have a benchmark for this?

**Comment by Mickeon** [comment]
Performance at the cost of binary size, which is worth evaluating since these methods are so common.

**Comment by Nazarwadim** [comment]
> Do you have a benchmark for this?

I tested the performance of my class compared to the previous one in c++.
This is the percentage increase to the methods:
normalized -> 19.2%
normalize -> 23.5%
angle -> 5%
from_angle -> 20%
length -> 15.5%
length_squared -> 29%
angle_to -> 25.5%
angle_to_point -> 3%
dot -> 28%
cross -> 33%
distance_to -> 23%
distance_squared_to -> 15%
sign -> 19.2%
floor -> 18.1%
ceil -> 20%
round -> 18.5%
rotated -> 17%
posmod -> 6.8%
posmodv -> 8.8%
project -> 25%
clamp -> 21%
snapped -> 18.1%
limit_length -> 16%
move_toward -> 5.8%
is_equal_approx -> 9%
is_zero_approx -> 9.4%
is_finite -> 6.3%

**Comment by Nazarwadim** [comment]
> Performance at the cost of binary size, which is worth evaluating since these methods are so common.

I checked it out and it turns out that my commit binary weighs 4kb less than the previous commit binary. IDK why, but ok
image

**Comment by Nazarwadim** [comment]
> What was the benchmark like? Was it microbenchmarking, like running the same function many times in a loop?

Yes, like this
image

I also tested 2D physics (spam riddibody) and its performance improved by 6.8% (tested several times)

https://github.com/godotengine/godot/assets/116561933/3a8a17df-54b5-4198-b962-8206f28afc02

**Comment by KoBeWi** [comment]
Probably worth checking if it makes difference with various compiler optimizations enabled. There's a chance compiler already does this automatically with the right flags.

**Comment by Chubercik** [comment]
Might also want to investigate potential benefits (if they really exist in official builds) in other vectors, i.e. `Vector3`, `Vector4`, and also `Vector2i`, `Vector3i` and `Vector4i`.

**Comment by aaronfranke** [comment]
@KoBeWi It's impossible to inline these at each file's compile time if the definition is in the `.cpp` file, because it's not available just from other files reading the header. It might be possible with LTO but I don't think compilers do this.

**Comment by Nazarwadim** [comment]
> Probably worth checking if it makes difference with various compiler optimizations enabled. There's a chance compiler already does this automatically with the right flags.

So, I wanted to write the same thing as @aaronfranke, but I decided to test it. 
And as it turned out, when compiling with `lto=full` mode, some built-in methods are inlined. But it's not clear which ones, so I built two projects, the first before my changes on commit 23191b834, the second on commit 8cfb6c8c4. Compiling time: 23191b834  9:15 8cfb6c8c4 9:19. The command to build is `scons lto=full`. 
The first thing I saw was that the build 8cfb6c8c4 weighed 73 KB less than 23191b834. I decided to test each method to which `_FORCE_INLINE_` was added via gdScript. I tested each method 100 times(and several additional times for methods with a large difference) to be more accurate. During the test, you need to turn off all applications (except for Godot) and do not move the mouse. 
Code:
```
func _ready() -> void:
	var sum_ms:int = 0
	for j in 100:
		var start := Time.get_ticks_msec()
		var length:int
		for i in 10000000:
			Vector2(i,i).”method”
		var end := Time.get_ticks_msec()
		sum_ms += end - start
		print("new version test:\n", end - start, " ms")
	print("avg <new/old> test time:", sum_ms/100.0)
```
Table

| methods             | "time(ms) in cpp file" | "time(ms) in header file" | "speed increase% ((cpp – header)/cpp * 100%)" |
| ---                 | ---                    | ---                       | ---                                           |
| angle               | 565.32                 | 560.74                    | 0.810                                         |
| angle_to            | 796.96                 | 792                       | 0.622                                         |
| angle_to_point      | 619.08                 | 607.34                    | 1.896                                         |
| clamp               | 608.56                 | 594.32                    | 2.339                                         |
| cross               | 587.2                  | 574.94                    | 2.087                                         |
| distance_squared_to | 586.12                 | 577.22                    | 1.518                                         |
| distance_to         | 573.16                 | 555.66                    | 3.053                                         |
| dot                 | 583.6                  | 577.6                     | 1.028                                         |
| floor               | 380.66                 | 380.96                    | -0.078                                        |
| from_angle          | 608.4                  | 591.2                     | 2.827                                         |
| is_equal_approx     | 377.2                  | 373.72                    | 0.922                                         |
| is_finite           | 362.66                 | 359.44                    | 0.887                                         |
| is_normalized       | 366.26                 | 365.28                    | 0.267                                         |
| is_zero_approx      | 363.44                 | 358.8                     | 1.276                                         |
| length              | 362.78                 | 363.24                    | -0.126                                        |
| length_squared      | 358.6                  | 357.5                     | 0.306                                         |
| limit_length        | 552.06                 | 547.44                    | 0.836                                         |
| move_toward         | 848.96                 | 839.76                    | 1.083                                         |
| normalize           | -----                  | --------------            | (not exist in gdscript)                       |
| normalized          | 370.24                 | 368.7                     | 0.415                                         |
| posmod              | 1150.54                | 1146.9                    | 0.316                                         |
| posmodv             | 969.02                 | 979.58                    | -1.089                                        |
| project             | 370.74                 | 370.86                    | -0.032                                        |
| round               | 384.28                 | 383.48                    | 0.2                                           |
| rotated             | 456.78                 | 449.14                    | 1.672                                         |
| sign                | 325.92                 | 323.32                    | 0.740                                         |
| snapped             | 429.26                 | 426.56                    | 0.628                                         |

File: core/math/vector2.h [hunk 1]
```diff
@@ -241,14 +241,14 @@ _FORCE_INLINE_ bool Vector2::operator!=(const Vector2 &p_vec2) const {
 	return x != p_vec2.x || y != p_vec2.y;
 }
 
-Vector2 Vector2::lerp(const Vector2 &p_to, const real_t p_weight) const {
+_FORCE_INLINE_ Vector2 Vector2::lerp(const Vector2 &p_to, const real_t p_weight) const {
```
Comment (lawnjelly) [inline core/math/vector2.h:244]: Are you sure you need the `_FORCE_INLINE_` in both the declaration and the definition? I haven't really tested this, but I see a lot of existing code just uses it in the declaration.

Comment (Chubercik) [reply core/math/vector2.h:244]: https://github.com/godotengine/godot/pull/87093#discussion_r1450150604

Seems like the best course of action would be to use inlining in the header file coupled with a definition and remove that method from the `.cpp` file.

**Comment by lawnjelly** [comment]
> I decided to test each method to which _FORCE_INLINE_ was added via gdScript.

I'm not sure that testing the effects of inlining in c++ is sensible via gdscript, which adds huge overheads. Profiling should show this.

File: core/math/vector2.h (hunk 1 above)
Comment (Nazarwadim) [reply core/math/vector2.h:244]: I removed `_FORCE_INLINE_` from the declaration and added it to the definition

**Comment by Nazarwadim** [comment]
> So I can see benefits to inlining these methods as it allows optimizations based on the operations, merging operations, etc.
> 
> However my concern is that these files are included almost everywhere, and while it might not hurt binary size the includes do mount up and affect compilation, especially on lower end processors, the engine is already quite heavy to build which I don't think will be helped by this
> 
> Edit: To give some perspective on it, `vector2.h` is included whenever you include either `object.h`, or `variant.h`, I can't give any kind of exact value but I'd say over 95% of the total codebase, so keeping these files lean is important

it may be worth moving such heavy methods as bezier_interpolate, cubic_interpolate_in_time, cubic_interpolate, bezier_derivative, slerp to cpp. 
I also see a lot of code in other classes with heavy methods that are implemented in `.h` even without the `_FORCE_INLINE` keyword

**Comment by AThousandShips** [comment]
I'd also say that a large portion of all files *including* this doesn't use any of it, *and* a lot of the code that uses is are in turn not inlined, such as mathematics in `Transform2D` etc., so there's still a lot of use cases that can't be really optimized well, the main use for optimization I'd say is when we have well defined input values that can be propagated, generally these methods are so short that inlining them doesn't change all that much unless I'm mistaken. So a proper profiling, and comparisons of build times and binary size would be necessary here, and I think doing this very selectively

**Comment by AThousandShips** [comment]
Another important thing to keep in mind is that any change to this header will cause the vast majority of the codebase to be rebuilt, we hopefully won't change these methods often, but that is an important consideration

Further, the angle methods use trigonometry, I'm not sure of the performance on modern processors but are those methods really so fast that a jump matters performance wise? Also a lot of this probably *can't* effectively be optimized in collapsing, folding, or moving operations, etc. so the benefit is limited

I'll compile this in some configurations and compare binary size to see, but can't help with profiling at the moment as I don't have anything set up

**Comment by lawnjelly** [comment]
I personally would agree with moving some of funcs to the headers (without much of a second thought), it's pretty much standard practice in my experience.

They probably should have been there from day 1, but I suspect what happens is that one contributor puts such a function in cpp, then other contributors copy.

In terms of justification, it is usually pretty obvious, and there is little need for profiling / benchmarking a priori. Of course sometimes profiling reveals something non-obvious which would benefit etc.

Benchmarking things like this is a bit of a fine art actually. There's a *lot* of opportunity to get false results (even for people who know what they are doing). Optimizer can optimize things out completely, you get cold / hot cache effects etc. There are some libraries / websites dedicated to making this kind of thing easier (e.g. google benchmarks, quickbench etc). Cache miss in particular can dwarf inlining.

There are of course trade offs such as:
* Runtime speed
* Compile time (potentially increased dependencies, and editing headers)
* Executable size

Some example guidelines:

### Indicators to put in header
* short function, likely to be bottleneck
* commonly called (i.e. hundreds / thousands of times per frame, rather than 1x per frame) - "hot functions"

### Indicators to be in cpp
* large functions
* seldom used - "cold functions"
* requires expensive header include (often better to put in the cpp, to prevent dependency creep and slow compile times)
* calls another expensive function (on a hot path) e.g. `print_line()` or does something slow (e.g. access disk)

`_FORCE_INLINE_` on the other hand is something we should take some care about imo, as it overrides compiler decisions in a lot of cases - it may be beneficial in some but not in others. Worth further reading on. It is kind of like saying "forget about your cost-benefit analysis and do as I say", although compiler can ignore it in e.g. recursive functions etc.

**Comment by Nazarwadim** [comment]
So I moved these methods to `.cpp`:
`slerp`, `cubic_interpolate`, `cubic_interpolate_in_time`, `bezier_interpolate`, `bezier_derivative`, `angle`, `from_angle`.

File: core/math/vector2.h [hunk 2]
```diff
@@ -314,6 +276,122 @@ _FORCE_INLINE_ Vector2 operator*(const int64_t p_scalar, const Vector2 &p_vec) {
...
+}
+
+_FORCE_INLINE_ Vector2 Vector2::normalized() const {
+	Vector2 v = *this;
+	v.normalize();
+	return v;
+}
+
+_FORCE_INLINE_ bool Vector2::is_normalized() const {
+	// use length_squared() instead of length() to avoid sqrt(), makes it more stringent.
```
Comment (AThousandShips) [inline core/math/vector2.h:303]:
```suggestion
	// Use length_squared() instead of length() to avoid sqrt(), makes it more stringent.
```
While we're here

File: core/math/vector2.h [hunk 3]
```diff
@@ -181,6 +159,31 @@ struct _NO_DISCARD_ Vector2 {
...
+	return x < y ? Vector2::AXIS_Y : Vector2::AXIS_X;
+}
+
+_FORCE_INLINE_ Vector2 Vector2::min(const Vector2 &p_vector2) const {
+	return Vector2(MIN(x, p_vector2.x), MIN(y, p_vector2.y));
+}
+
+_FORCE_INLINE_ Vector2 Vector2::max(const Vector2 &p_vector2) const {
+	return Vector2(MAX(x, p_vector2.x), MAX(y, p_vector2.y));
+}
```
Comment (AThousandShips) [inline core/math/vector2.h:185]: I'd restore these to their original place, less cluttered IMO, it also matches the other classes which matters

File: core/math/vector2.h [hunk 4]
```diff
@@ -181,6 +159,31 @@ struct _NO_DISCARD_ Vector2 {
 	}
 };
 
+_FORCE_INLINE_ real_t &Vector2::operator[](int p_idx) {
+	DEV_ASSERT((unsigned int)p_idx < 2);
+	return coord[p_idx];
+}
+_FORCE_INLINE_ const real_t &Vector2::operator[](int p_idx) const {
```
Comment (AThousandShips) [inline core/math/vector2.h:166]:
```suggestion

_FORCE_INLINE_ const real_t &Vector2::operator[](int p_idx) const {
```
If we keep them here, leave a line between each implementation

File: core/math/vector2.cpp [hunk 5]
```diff
@@ -33,48 +33,53 @@
 #include "core/math/vector2i.h"
 #include "core/string/ustring.h"
 
-real_t Vector2::angle() const {
```
Comment (AThousandShips) [inline core/math/vector2.cpp:36]: Please put the methods you've moved at the end to reduce the noise, the methods that haven't been changed shouldn't be in the diff

File: scene/gui/tree.cpp [hunk 6]
```diff
@@ -120,6 +120,10 @@ void TreeItem::_change_tree(Tree *p_tree) {
 		}
 
 		if (tree->selected_item == this) {
+			for (int i = 0; i < tree->selected_item->cells.size(); i++) {
+				tree->selected_item->cells.write[i].selected = false;
+			}
+
```
Comment (Mickeon) [inline scene/gui/tree.cpp:126]: You have committed this unrelated change in this PR. I don't think it was intentional?

Comment (Nazarwadim) [reply scene/gui/tree.cpp:126]: you can see from the commit history that I haven't learned how to use rebase yet)

**Comment by Mickeon** [comment]
By the way, I hope you do not take the core team's skepticism at heart. They just want to ensure this is an objectively good change because this is a prickly part of the engine.

**Comment by Nazarwadim** [comment]
> By the way, I hope you do not take the core team's skepticism at heart. They just want to ensure this is an objectively good change because this is a prickly part of the engine.

In theory, I've moved all the heavy methods into `cpp` and they shouldn't impact performance. And the methods that I have implemented in the header improve performance because they aren`t complex. But just in case, someone else should check the binary size and performance. Maybe I made a mistake somewhere.

**Comment by Nazarwadim** [comment]
So, in general, I moved the following methods to `.cpp`:
* `plane_project` is used 2 times in the project.
* `slerp` `cubic_interpolate` `cubic_interpolate_in_time` `bezier_interpolate` `bezier_derivative` are very complex and are almost not affected by `_FORCE_INLINE_`.

Methods that I moved to `.h`:
* `length` `length_squared` `distance_to` `distance_squared_to` `dot` `cross` `sign` `floor` `ceiling` `round` `is_equal_approx` `is_zero_approx` `is_finite` simple one-line methods that are often used throughout the code base. I placed these methods in the body of the class.
* `normalize` `normalized` `is_normalized` `clamp` - methods that are also simple and used frequently, but I placed outside the class body.

Additional work I have done:
* Reused `length_squared` and `distance_squared_to`.
* Removed `_FORCE_INLINE_` from methods that are declared but not implemented. Added `_FORCE_INLINE_` to the defined methods.
* `// Use length_squared() instead of length() to avoid sqrt(), makes it more stringent.` replaced `use` with `Use`.
* Shortened the methods `cubic_interpolate` `cubic_interpolate_in_time` `bezier_interpolate` `bezier_derivative`.

**Review by Chubercik — APPROVED** [review]

**Comment by Mickeon** [comment]
How do we feel about the PR as is now?

**Comment by Chubercik** [comment]
> How do we feel about the PR as is now?

Given that this PR is all about performance, maybe some [godot-benchmarks](https://github.com/godotengine/godot-benchmarks) tests are in order.

**Comment by Calinou** [comment]
> Given that this PR is all about performance, maybe some [godot-benchmarks](https://github.com/godotengine/godot-benchmarks) tests are in order.

Note that godot-benchmarks tests all use the scripting API (GDScript/C#/GDExtension), so if your goal is to look at pure C++ performance, it's not a suitable project.

**Review by Calinou — COMMENTED** [review]
Tested locally, it works as expected.

Binary size of a stripped Linux x86_64 optimized editor binary (`optimize=speed lto=full`) decreases by 110 KB with this PR:

```
122,973,808  godot.linuxbsd.editor.x86_64.master
122,863,184  godot.linuxbsd.editor.x86_64
```

This is counterintuitive since I'd expect forcing inlining to *increase* binary size.

What's also counterintuitive is that performance is pretty much identical to before in this testing project: [test_vector2_performance.zip](https://github.com/user-attachments/files/16281054/test_vector2_performance.zip)

```
angle: 311 ms
angle_to: 230 ms
angle_to_point: 332 ms
ceil: 205 ms
clamp: 211 ms
cross: 205 ms
distance_squared_to: 213 ms
distance_to: 212 ms
dot: 213 ms
floor: 208 ms
from_angle: 230 ms
is_equal_approx: 210 ms
is_finite: 209 ms
is_normalized: 208 ms
is_zero_approx: 212 ms
length: 211 ms
length_squared: 209 ms
limit_length: 310 ms
move_toward: 230 ms
normalized: 213 ms
posmod: 738 ms
posmodv: 724 ms
project: 212 ms
round: 235 ms
rotated: 228 ms
sign: 216 ms
```

The figures I get before and after this PR are within 2% of each other. Something I noticed is that `posmod()` and `posmodv()` are quite a bit slower than every other method out there – is there room for optimization here?

<details open>

<summary>PC specifications</summary>

- **CPU:** Intel Core i9-13900K
- **GPU:** NVIDIA GeForce RTX 4090
- **RAM:** 64 GB (2×32 GB DDR5-5800 C30)
- **SSD:** Solidigm P44 Pro 2 TB
- **OS:** Linux (Fedora 40)
</details>

File: core/math/vector2.h [hunk 7]
```diff
@@ -81,56 +81,58 @@ struct [[nodiscard]] Vector2 {
...
-	Vector2 maxf(real_t p_scalar) const {
+	_FORCE_INLINE_ Vector2 maxf(real_t p_scalar) const {
 		return Vector2(MAX(x, p_scalar), MAX(y, p_scalar));
 	}
 
-	real_t distance_to(const Vector2 &p_vector2) const;
-	real_t distance_squared_to(const Vector2 &p_vector2) const;
+	_FORCE_INLINE_ real_t distance_to(const Vector2 &p_vector2) const { return Math::sqrt(distance_squared_to(p_vector2)); }
+	_FORCE_INLINE_ real_t distance_squared_to(const Vector2 &p_vector2) const {
+		return (x - p_vector2.x) * (x - p_vector2.x) + (y - p_vector2.y) * (y - p_vector2.y);
```
Comment (Chaosus) [inline core/math/vector2.h:106]:
```suggestion
		real_t x2 = x - p_vector2.x;
		real_t y2 = y - p_vector2.y;
		return x2 * x2 + y2 * y2;
```

I think it's a good optimization for this function.

Comment (Chubercik) [reply core/math/vector2.h:106]: How is it an optimization? Looks better, but best case scenario does the same thing.

Comment (Chaosus) [reply core/math/vector2.h:106]: It generates less assembly instructions:
24 vs 31

Comment (Chubercik) [reply core/math/vector2.h:106]: Not sure where you got that, with the optimization flag enabled it produces the same instructions:
https://godbolt.org/z/xx9ob8dGb

Comment (Chaosus) [reply core/math/vector2.h:106]: Nevermind..

Comment (Chubercik) [reply core/math/vector2.h:106]: For the record, I'm not against this change, but it'd be more for clarity's sake, not performance.

**Comment by Nazarwadim** [comment]
I recently read that LTO inline functions.
I compiled `disable_3d=yes disable_advanced_gui=yes target=template_release lto=full` and the binary size was 200kb larger compared to master. I also did some benchmarks in c++ and there was almost no difference.

As for `_FORCE_INLINE_`, I can say that 99% of the time it is not necessary and can even do worse by greatly increasing the size of the binary and even reduce performance due to more CPU cache usage for the same instructions.

So, how can I conclude. If a method is normally declared in `.h` and implemented in `.cpp` and LTO is used, it is the same as if the method were implemented in `.h` and with the `inline` keyword. If `_FORCE_INLINE_` is used, then the method must be implemented in the header.
